### PR TITLE
PLU-347: feat(box): add ACL permissions metadata to Box connector

### PR DESCRIPTION
**Pull request by danielle-unstructured-io**
## Summary

- Extends the Box connector to populate `permissions_data` on `FileDataSourceMetadata`, consistent with the Confluence and Google Drive implementations
- Permissions are fetched at **index time** in `BoxIndexer.run()` so they're available to all downstream pipeline stages
- `BoxDownloader.run()` retains a fallback for standalone usage (CLI, integration tests without the SND plugin layer)

## What changed

**`unstructured_ingest/processes/connectors/fsspec/box.py`**
- Added `BOX_ROLE_MAPPING` — maps Box collaboration roles to `[read]`, `[read, update]`, or `[read, update, delete]`; `uploader` excluded (write-only)
- Added module-level `_normalize_collaborations()`, `_get_collaborations_for_folder()`, `_get_permissions_for_file()` helpers
- `BoxIndexer.run()` override — initializes a Box SDK client once, then for each indexed file walks `path_collection` ancestor folders (LRU-cached, max 5 entries) plus direct file collaborations to build normalized permissions
- `BoxDownloader.run()` fallback — only fetches permissions if `permissions_data is None` (i.e., indexer wasn't run)
- `BoxConnectionConfig.get_box_client()` — returns an authenticated `boxsdk.Client` via JWT

**Tests**
- 21 unit tests covering `BOX_ROLE_MAPPING`, `_normalize_collaborations`, and `_get_permissions_for_file` (all mock-based)
- Integration test + expected-results fixtures for both top-folder and second-tier subfolder files

## Test plan

- [x] All 21 unit tests pass
- [x] Verified end-to-end in SND: root-level and subfolder Box files both have `permissions_data` populated with correct user IDs
- [ ] Integration tests pass in CI

Closes PLU-347

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adds new Box API calls during indexing/downloading to derive and attach normalized ACL metadata, which can impact performance and correctness of permission propagation if edge cases are missed. Risk is mitigated by caps, caching, and extensive unit/integration test coverage.
> 
> **Overview**
> Adds **ACL permission metadata passthrough** to the Box connector by deriving `metadata.permissions_data` from Box collaborations and normalizing them into `read`/`update`/`delete` user/group ID lists.
> 
> Permissions are fetched primarily at **index time** via an ancestor-folder collaboration walk with a small LRU cache and a configurable cap (`BoxIndexerConfig.max_num_metadata_permissions`), with a downloader-side fallback for standalone usage.
> 
> Updates/extends tests: new Box unit tests for role mapping and permission normalization, new Box integration tests plus expected fixtures, and a test harness fix to strip randomized `unstructured_<tmp>/` path segments so fixture directory structures are stable. Docs images paths and package version/changelog are also updated for the `1.6.0` release.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 312f7c53adadb227a8ed623fec6d6a42a95916d2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: unstructured_ingest/processes/connectors/fsspec/box.py [hunk 1]
```diff
@@ -33,12 +35,105 @@
...
+        entity_id = accessible_by.get("id")
+        if not entity_id:
+            continue
+        operations = BOX_ROLE_MAPPING.get(collab.get("role", ""), [])
+        if not operations:
+            continue
+        type_key = entity_type + "s"
+        for op in operations:
+            normalized[op][type_key].add(entity_id)
+        total[0] += 1
```
[the remark this answers is not in the document]
Comment (awalker4) [reply unstructured_ingest/processes/connectors/fsspec/box.py:84]: I'd have claude take a look at this one. (This being the Cursor comment about overgranted permissions)

**Review by awalker4 — COMMENTED** [review]
Will just need to add a new block to the changelog for whatever the next version is. This value also goes into `__version__.py`

**Comment by awalker4** [comment]
@claude please review. Be thorough when looking at the ACL properties as we want to make sure this is correct

File: unstructured_ingest/processes/connectors/fsspec/box.py [hunk 2]
```diff
@@ -95,6 +211,15 @@ def wrap_error(self, e: Exception) -> Exception:
         logger.error(f"unhandled exception from box ({type(e)}): {e}", exc_info=True)
         return e
 
+    @requires_dependencies(["boxsdk"], extras="box")
```
Comment (brandon-u10d) [inline unstructured_ingest/processes/connectors/fsspec/box.py:247]: `boxsdk` is not listed in the `box` extras in `pyproject.toml` (`box = ["boxfs", "fsspec"]`). This decorator tells users to `pip install unstructured-ingest[box]`, but that won't install `boxsdk`. Confirmed `boxfs` does not pull in `boxsdk` as a transitive dependency either.

The ACL feature silently degrades (the `except` at line 268 catches the import failure and logs a warning, yielding `None` permissions). One-line fix: add `"boxsdk"` to the extras in `pyproject.toml`.

File: unstructured_ingest/processes/connectors/fsspec/box.py [hunk 3]
```diff
@@ -146,6 +294,26 @@ class BoxDownloader(FsspecDownloader):
...
+        # permissions_data is set during indexing in BoxIndexer.run(); this fallback handles
+        # standalone downloader usage (e.g., CLI, integration tests without the SND plugin layer).
+        # Memoize client and cache across files so we don't re-run JWT auth per file.
+        if file_data.metadata.permissions_data is None:
+            file_id = (file_data.metadata.record_locator or {}).get("file_id")
+            if file_id:
+                try:
+                    if self._box_client is None:
+                        self._box_client = self.connection_config.get_box_client()
+                    file_data.metadata.permissions_data = _get_permissions_for_file(
```
Comment (brandon-u10d) [inline unstructured_ingest/processes/connectors/fsspec/box.py:367]: The downloader fallback calls `_get_permissions_for_file` without passing `max_perms`, falling through to the default of 500. The indexer at line 280 passes `self.index_config.max_num_metadata_permissions`. If a user configures a non-default cap (say 50), the indexer respects it but the standalone downloader path silently ignores it.

Confluence avoids this by putting `max_num_metadata_permissions` on `ConfluenceDownloaderConfig` (confluence.py:249). Worth mirroring that pattern on `BoxDownloaderConfig` if the standalone path is a supported use case.

File: unstructured_ingest/processes/connectors/fsspec/box.py [hunk 4]
```diff
@@ -33,14 +35,128 @@
...
+        ):
+            collabs.append(collab.response_object)
+    except Exception as e:
+        # Don't cache on failure — a transient 503/timeout would otherwise silently
+        # zero out permissions for every descendant file sharing this ancestor for
+        # the rest of the indexer run.
+        logger.warning(f"Could not retrieve collaborations for folder {folder_id}: {e}")
+        return collabs
+
+    if len(cache) >= 5:
```
Comment (brandon-u10d) [inline unstructured_ingest/processes/connectors/fsspec/box.py]: Cache capacity is hardcoded to 5. Confluence makes this configurable via `_permissions_cache_max_size: int = 5` (confluence.py:277). For Box, the cache stores ancestor *folder* collaborations, and folder trees with more than 5 sibling folders at any level will thrash (evict entries needed by the next file in the same branch). Worth exposing on `BoxIndexerConfig` alongside `max_num_metadata_permissions`, with a higher default (e.g. 128) since each entry is a small list of dicts.

File: unstructured_ingest/processes/connectors/fsspec/box.py [hunk 5]
```diff
@@ -33,14 +35,128 @@
...
+        "read": {"users": set(), "groups": set()},
+        "update": {"users": set(), "groups": set()},
+        "delete": {"users": set(), "groups": set()},
+    }
+    total = [0]
+
+    # Default True so a failed file.get() doesn't suppress the direct-collab fetch.
+    has_direct_collabs = True
+    try:
+        file_obj = client.file(file_id).get(fields=["path_collection", "has_collaborations"])
```
Comment (brandon-u10d) [inline unstructured_ingest/processes/connectors/fsspec/box.py]: Every file calls `client.file(file_id).get(fields=["path_collection", "has_collaborations"])` to discover its ancestor chain, but files sharing a parent folder return identical `path_collection` data. For 1,000 files across 50 folders, that's ~950 redundant calls.

A follow-up optimization: cache the path-to-ancestors mapping per parent folder via `folder.get(fields=["path_collection"])` once per unique parent, then reuse for all files within it. Not blocking for the initial ship, but worth tracking since Box rate-limits at 1,000 calls/minute for Enterprise accounts.

File: unstructured_ingest/processes/connectors/fsspec/box.py [hunk 6]
```diff
@@ -33,14 +36,160 @@
...
+    # given folder — material savings against Box's 1k-calls/minute Enterprise rate limit.
+    # Default True so a failed file.get() doesn't suppress the direct-collab fetch.
+    has_direct_collabs = True
+    ancestor_folder_ids: Optional[list[str]] = None
+    if (
+        parent_chain_cache is not None
+        and parent_path is not None
+        and parent_path in parent_chain_cache
+    ):
+        ancestor_folder_ids = parent_chain_cache[parent_path]
```
Comment (brandon-u10d) [inline unstructured_ingest/processes/connectors/fsspec/box.py:141]: The `parent_chain_cache` hit skips `file.get()` (good), but also skips reading `has_collaborations`, so `has_direct_collabs` stays at its default `True`. Sibling files in folders without direct collaborations make one unnecessary `get_collaborations()` call each.

Non-blocking; matches every other connector's unconditional-fetch pattern. Small follow-up optimization:

```suggestion
        ancestor_folder_ids, has_direct_collabs = parent_chain_cache[parent_path]
```

Corresponding write at line 155: `parent_chain_cache[parent_path] = (ancestor_folder_ids, has_direct_collabs)`

**Review by brandon-u10d — APPROVED** [review]
All four prior comments verified addressed.
